### PR TITLE
Library overhauls

### LINKED ISSUE #36: Favor giving names to functions over operators
It might be nice to provide named functions for interfaces first, then have the operators as shorcuts. So for example, `(>>=)` would become `flat_map` (I personally prefer this over `bind`, heh), and `make_Monad` would define the operator based on that.

### PR DESCRIPTION
**Pull request by bjz**
Fixes #36

This is a huge overhaul of the prelude API. It removes duplication, and give names to most of the more obscure operators. I've also enforced interface inheritance using fields, for example:

```diff
 type Applicative f = {
-    (<*>) : f (a -> b) -> f a -> f b,
+    functor : Functor f,
+    apply : f (a -> b) -> f a -> f b,
     pure : a -> f a
 }
```

This allows us to remove some of the parameters from the `make_*` functions, and is more self-documenting. It is also a nice step along the road to modular implicits (see #47). Despite my concerns (https://github.com/Marwes/gluon/issues/47#issuecomment-233562336), using fields doesn't seem too onerous, at least at this stage. It was neat to see that the language as it stands already allowed us to do this! 😄 

I have a feeling that the new API is beginning to hurt compile times, but I would rather we tackle that in a separate pass. There are already some PRs out about this (#50, #41, #11), but it might also pay to start profiling it to see if there are any places we can improve things.

I'm guessing the continual calling of the `make_*` functions is causing lots of allocations. Maybe the VM could be smarter about this? If we had a better effect system, perhaps we could figure out what is constant, and cache values based on their input args.... I dunno. Probably needs to be a separate issue. And as always, speculating on performance is usually pretty ineffective without profiling!

### CODE REVIEW

File: src/io.rs [hunk 1]
```diff
@@ -204,7 +204,7 @@ pub fn load(vm: &Thread) -> Result<()> {
     vm.add_bytecode("io_bind", io_bind_type, 3, io_bind);
```
Comment (bjz) [inline src/io.rs]: I would like to rename this to `io_flat_map`, but it wasn't very clear to me how to rewrite it with the arguments swapped. Still haven't got my head around the VM's bytecode! :o

File: std/prelude.glu [hunk 2]
```diff
@@ -322,169 +361,169 @@ let functor_Result: Functor (Result e) = {
...
-    let { (>>=), return } = m
+    let { applicative, flat_map } = m
+    let { pure } = applicative
     let { id } = category_Function
 
-    let (>>) l r = l >>= \_ -> r
-    let forM_ xs f =
+    let (=<<) : (a -> m b) -> m a -> m b = flat_map
+    let (>>=) : m a -> (a -> m b) -> m b = flip flat_map
+    let (>>) l r : m a -> m b -> m b = l >>= \_ -> r
```
Comment (bjz) [inline std/prelude.glu]: Still need to remove this in favor of `(*>)`

File: src/io.rs (hunk 1 above)
Comment (Marwes) [inline src/io.rs]: I believe this should work (untested)
```
// Stack
//  [f, m, ()]
let io_bind = vec![Call(1), // [f, m_ret] // Call m ()
                   PushInt(0), // [f, m_ret, ()] // Add a dummy argument ()
                   TailCall(2)]; // [f_ret] // Call f m_ret ()

```
IO actions are implement as simply having an extra argument (`IO a = () -> a` so invoking an action just needs to add an extra dummy argument and calling the function.

I can add some documentation to the bytecode though if you want, currently the documentation is just "what the vm does".

Comment (bjz) [inline src/io.rs]: That would be amazing! I reckon the right place would be on the variants.

Comment (bjz) [inline src/io.rs]: That worked as expected 👍

**Comment by Marwes** [comment]
> Despite my concerns (#47 (comment)), using fields doesn't seem too onerous, at least at this stage...

This looks good indeed. One way to otherwise get some niceness back if we feel it is necessary would be to add the fields of an "inherited record" to the result of the `make_*` functions.

> I have a feeling that the new API is beginning to hurt compile times, but I would rather we tackle that in a separate pass.

Definitely need to properly profile this as I don't have a good idea of where the time is spent at the moment. Wonder if there is a good way to setup automatic ratcheting of benchmarks with travis so that it is possible to track performance regressions automatically (travis might have to variable times though).

> I'm guessing the continual calling of the make_* functions is causing lots of allocations...

Yep. A sufficiently smart compiler would be able to remove the allocations but the current compiler is as dumb as they come :). Actually implementing the necessary those optimizations may take some time (and I don't see gluon as a language which should be doing much optimizations, at least at the moment). I have an idea for how we may provide a workaround in the meantime though with macros (I will post an issue in a bit).

**Comment by bjz** [comment]
Agreed about being conservative about optimisations. Less magic would be great, if possible.

**Comment by bjz** [comment]
Speak of the devil... https://travis-ci.org/Marwes/gluon/jobs/146827271#L632

> thread 'tutorial_13' panicked at 'command failed', /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/skeptic-0.4.0/lib.rs:308

Retrying the build...

**Comment by bjz** [comment]
Succeeded on nightly and stable though. Probably a spurious failure.

**Comment by Marwes** [comment]
And I have already restarted that build once...

**Comment by bjz** [comment]
Seems to pass now.

**Comment by bjz** [comment]
Rebuilt based on the latest master - is there anything that you had issues with? Sorry for it being such a large PR - got a little carried away :(

**Comment by Marwes** [comment]
Nope, looks good to me!

**Pull request merged**